docs/proposals/20210301-enhancement_of_YurtHub_caching_ability.md

```diff
@@ -0,0 +1,257 @@
+---
+title: Proposal to enhance the caching ability of YurtHub
+authors:
+  - "@qclc"
+  - "@GsssC"
+reviewers:
+  - "@rambohe-ch"
+creation-date: 2021-03-01
+last-updated: 2021-03-01
+status: provisional
+---
+
+# Proposal to enhance the caching ability of YurtHub
+
+## Table of Contents
+
+- [Title](#title)
+  - [Table of Contents](#table-of-contents)
+  - [Summary](#summary)
+  - [Motivation](#motivation)
+    - [Goals](#goals)
+  - [Description of Yurthub's existing cache ability](#Description of Yurthub's existing cache ability)
+    - [serializerManager](#serializerManager) 
+      - [Problem](#Problem) 
+    - [cacheManager](#cacheManager) 
+      - [Problem](#Problem) 
+  - [Proposal](#proposal)
+    - [Delete resourceToKindMap and resourceToListKindMap](#Delete resourceToKindMap and resourceToListKindMap)
+      - [Design detail](#Design detail)
+    - [Decode the unregistered Custom Resources into Unstructured structure](#Decode the unregistered Custom Resources into Unstructured structure)
+      - [Design detail](#Design detail)
+        - [Add unstructuredNegotiatedSerializer](#Add unstructuredNegotiatedSerializer)
+        - [Determine the resource type based on request.URL and select the appropriate serializer](#Determine the resource type based on request.URL and select the appropriate serializer)
+  - [Implementation History](#implementation-history)
+
+## Summary
+
+YurtHub can currently only cache resources defined in  [resourceToKindMap](https://github.com/openyurtio/openyurt/blob/4d7463a40801c29d09c4f7d10ba46b73cb019915/pkg/yurthub/cachemanager/cache_manager.go#L46) and [resourceToListKindMap](https://github.com/openyurtio/openyurt/blob/4d7463a40801c29d09c4f7d10ba46b73cb019915/pkg/yurthub/cachemanager/cache_manager.go#L62). This proposal mainly tries to put forward and realize a set of scheme, so that YurtHub is no longer limited by resource types, and Custom Resources can be cached.
+
+## Motivation
+
+when network between cloud and edge disconnected, if any pod(eg: calico) on the edge node that used some resources(like crd) not in the above map want to run continuously, that is to say, the pod can not restarted successfully because resources(like crd) are not cached by yurt-hub.
+
+### Goals
+
+- Delete [resourceToKindMap](https://github.com/openyurtio/openyurt/blob/4d7463a40801c29d09c4f7d10ba46b73cb019915/pkg/yurthub/cachemanager/cache_manager.go#L46) and [resourceToListKindMap](https://github.com/openyurtio/openyurt/blob/4d7463a40801c29d09c4f7d10ba46b73cb019915/pkg/yurthub/cachemanager/cache_manager.go#L62) so that we can remove the restriction that yurthub can only cache certain resources;
+- Decode Custom Resources that are not registered in the scheme into `Unstructured` structures;
+- The Custom Resource is distinguished from the core resource based on `request.URL`. The core resource uses the existing serializer and the Custom Resource uses the `unstructuredNegotiatedSerializer`.
```

Comment:
How about use `unstructuredNegotiatedSerializer` for all of resources?

Reply:
Unstructured objects still have functioning TypeMeta features-- kind, version, etc, but for private properties that distinguish resources, such as specs, registered structures are required to parse them out in detail. If use unstructuredNegotiatedSerializer to decode all resource, the Yurthub will lose the flexibility to operate core resources.
Or if you mean to use only one NegotiatedSerializer to handle all the resources, our current idea is to later encapsulate the two kind of serializers into one and decide internally which one to use.

Reply:
> Unstructured objects still have functioning TypeMeta features-- kind, version, etc, but for private properties that distinguish resources, such as specs, registered structures are required to parse them out in detail. If use unstructuredNegotiatedSerializer to decode all resource, the Yurthub will lose the flexibility to operate core resources.
> Or if you mean to use only one NegotiatedSerializer to handle all the resources, our current idea is to later encapsulate the two kind of serializers into one and decide internally which one to use.

For the sake of flexibility to operate resources, we should use `unstructuredNegotiatedSerializer` uniformly, and then use  `NegotiatedSerializer ` only when resources need to be considered for detail.